File: src/types/progress.ts
export interface ProfilePath {
  id: number;
  title: string;
}

export interface ProfileSkill {
  name: string;
  level: number;
  xp: number;
}

export interface Activity {
  id: number;
  type: string;
  description: string;
  xp_earned: number;
  timestamp: string | null;
}

export interface UserProfile {
  id: number;
  display_name: string;
  avatar_seed: string;
  level: number;
  title: string;
  total_xp: number;
  xp_to_next_level: number;
  quests_completed: number;
  current_path: ProfilePath | null;
  skills: ProfileSkill[];
  recent_activity: Activity[];
}

export interface ProfileAchievement {
  id: number;
  title: string;
  description: string | null;
  icon: string | null;
  category: string | null;
  unlocked: boolean;
  unlocked_at: string | null;
}

export interface ProgressSummary {
  total_lessons_completed: number;
  total_courses_completed: number;
  total_xp: number;
  current_level: number;
  current_level_xp: number;
  xp_to_next_level: number;
  level_progress_percentage: number;
  streak_days: number;
}

export interface RoadmapProgress {
  path_id: number;
  completed_nodes: number;
  total_nodes: number;
  completion_percentage: number;
}

export interface CourseUserProgress {
  completed_lessons: number;
  completion_percentage: number;
  started_at: string | null;
  last_accessed_at: string | null;
  completed_at: string | null;
}

export interface CourseProgressLesson {
  lesson_id: number;
  title: string;
  completed: boolean;
  xp_reward: number;
}

export interface CourseProgressDetail {
  course_id: number;
  course_title: string;
  completed_lessons: number;
  total_lessons: number;
  completion_percentage: number;
  total_xp: number;
  lessons: CourseProgressLesson[];
}

export interface AchievementUnlock {
  id: number;
  title: string;
  description: string | null;
  icon: string | null;
  category: string | null;
}

export interface CompleteLessonRequest {
  lesson_id: number;
  course_id: number;
  time_spent_seconds: number;
}

export interface CompleteLessonResponse {
  xp_earned: number;
  total_xp: number;
  level_before: number;
  level_after: number;
  xp_to_next_level: number;
  new_achievements: AchievementUnlock[];
  node_completed: boolean;
}

export interface UpdateProfileRequest {
  display_name?: string;
  avatar_seed?: string;
  current_path_id?: number;
}

File: server/schemas/progress.py
"""Live single-user profile and progress schemas."""

from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class CurrentPathResponse(BaseModel):
    id: int
    title: str


class SkillResponse(BaseModel):
    name: str
    level: int
    xp: int


class ActivityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    type: str
    description: str
    xp_earned: int
    timestamp: Optional[str] = None


class ProfileResponse(BaseModel):
    id: int
    display_name: str
    avatar_seed: str
    level: int
    title: str
    total_xp: int
    xp_to_next_level: int
    quests_completed: int
    current_path: Optional[CurrentPathResponse] = None
    skills: List[SkillResponse] = Field(default_factory=list)
    recent_activity: List[ActivityResponse] = Field(default_factory=list)


class ProfileUpdateRequest(BaseModel):
    display_name: Optional[str] = None
    avatar_seed: Optional[str] = None
    current_path_id: Optional[int] = None


class ProfileUpdateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    display_name: Optional[str] = None
    avatar_seed: str
    total_xp: int
    level: int
    current_path_id: Optional[int] = None


class AchievementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: Optional[str] = None
    icon: Optional[str] = None
    category: Optional[str] = None
    unlocked: bool
    unlocked_at: Optional[str] = None


class AchievementUnlockResponse(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    icon: Optional[str] = None
    category: Optional[str] = None


class SkillsListResponse(BaseModel):
    skills: List[SkillResponse] = Field(default_factory=list)


class ActivityListResponse(BaseModel):
    activities: List[ActivityResponse] = Field(default_factory=list)


class CompleteLessonRequest(BaseModel):
    lesson_id: int
    course_id: int
    time_spent_seconds: int = 0


class CompleteLessonResponse(BaseModel):
    xp_earned: int
    total_xp: int
    level_before: int
    level_after: int
    xp_to_next_level: int
    new_achievements: List[AchievementUnlockResponse] = Field(default_factory=list)
    node_completed: bool


class ProgressSummaryResponse(BaseModel):
    total_lessons_completed: int
    total_courses_completed: int
    total_xp: int
    current_level: int
    current_level_xp: int
    xp_to_next_level: int
    level_progress_percentage: int
    streak_days: int


class RoadmapProgressResponse(BaseModel):
    path_id: int
    completed_nodes: int
    total_nodes: int
    completion_percentage: float


class LessonProgressItem(BaseModel):
    lesson_id: int
    title: str
    completed: bool
    xp_reward: int


class CourseProgressResponse(BaseModel):
    course_id: int
    course_title: str
    completed_lessons: int
    total_lessons: int
    completion_percentage: float
    total_xp: int
    lessons: List[LessonProgressItem] = Field(default_factory=list)
